common/compositor/gl/glprogram.cpp

```diff
@@ -126,8 +126,93 @@ static std::string GenerateFragmentShader(int layer_count) {
   return fragment_shader_stream.str();
 }
 
+static GLint LoadPreBuiltBinary(GLint gl_program,
+                                void *binary, long size) {
+  GLint status;
+
+  std::string extensions((const char *)glGetString(GL_EXTENSIONS));
+  std::string mesa_prog_bin_fmts("GL_MESA_program_binary_formats");
+  std::string::size_type pos;
+
+  /* check if glProgramBinaryOES exists */
+  if (!glProgramBinaryOES)
+    return 0;
+
+  pos = extensions.find(mesa_prog_bin_fmts);
+
+  /* check GL_MESA_program_binary_formats exists */
+  if (pos == std::string::npos)
+    return 0;
+
+  /* TODO: we should check whether GL_MESA_program_binary_formats are
+   * supported here: however, this extension is not currently exposed now
+   * in MESA tip. Once it's fixed we will add a check routine here
+   */
+  glProgramBinaryOES(gl_program, GL_PROGRAM_BINARY_FORMAT_MESA, binary,
+                     size);
+
+  glGetProgramiv(gl_program, GL_LINK_STATUS, &status);
+  if (status)
+      return gl_program;
+
+  return 0;
+}
+
 static GLint GenerateProgram(unsigned num_textures,
                              std::ostringstream *shader_log) {
+  GLint status;
+  GLint program = glCreateProgram();
+  void *binary_prog;
+  long binary_sz;
+
+  if (!program) {
+    if (shader_log)
+      *shader_log << "Failed to create program."
+                  << "\n";
+    return 0;
+  }
+
+  /* try to load prebuilt shader program from files */
+  std::ostringstream shader_program_fname;
+  shader_program_fname << PREBUILT_SHADER_FILE_PATH"hwc_shader_prog_"
+                       << num_textures
+                       <<".shader_test.bin";
+
+  FILE *shader_prog_fp;
+
+  shader_prog_fp = fopen(shader_program_fname.str().c_str(), "rb");
+
+  if (shader_prog_fp) {
+    /* check the size of file */
+    fseek(shader_prog_fp, 0, SEEK_END);
+    binary_sz = ftell(shader_prog_fp);
+    rewind(shader_prog_fp);
+
+    char *binary_prog = (char *)malloc(binary_sz);
+    if (binary_prog) {
+      if (fread(binary_prog, 1, binary_sz, shader_prog_fp) == binary_sz) {
```

Comment:
Given that we generate shaders on-demand, I don't think it is such a great idea to block on disk access. @kalyankondapally have you considered generating all shaders in bulk during initialization?

Reply:
Kalyan and I just talked about this off-line, in the long run we are going to want to do bulk shader generation as close to program init as possible. For now you can keep this here and leave the feature disabled by default.

Reply:
@strassek : I see your request is still outstanding. Can you approve this if you are ok with this part now?

Reply:
I don't think it is a good idea to try to do this read by default, so I'd like to put this behind a build flag and have it disabled by default. Once we start doing bulk shader generation at init, then we can get rid of the build flag.

Reply:
I see. I will define a temporary build flag.